### PR TITLE
Callback function for each Scraper Execution, passed as second argument of enableContentGrabber()

### PR DESCRIPTION
**Pull request by dstuecken**
Hi all,

I wanted to easily be able to do something with the page content that the Scaper holds if the content grabber is enabled, so i added a getGrabber function to the Parser.

### CODE REVIEW

**Comment by dstuecken** [comment]
Thought about it again and found it more useful to define an optional callback function that is called for each scraper execution. Feed, Item and Scraper instances are passed into it.

**Comment by BernhardPosselt** [comment]
Can't review because I've done too little on the grabber, @fguillot @mkresin ?

**Comment by dstuecken** [comment]
One practical example on using this is:

```
$parser->enableContentGrabber(false, function(Feed $feed, Item $item, Scraper $scraper) {
    // Retrieve the first jpeg, gif or png image from content and override the possible existing item's enclosure_url
    if (preg_match('/\.*?src\=\"(.*\.(?:jpg|jpeg|gif|png))\"/i', $scraper->getRelevantContent(), $matches))
    {
        $item->enclosure_url = $matches[1];
    }
});
```

**Comment by BernhardPosselt** [comment]
So it's basically an after success callback for the content grabber?

**Comment by dstuecken** [comment]
Don't have to be a success. It is always executed after the content grabber has finished parsing the site.

**Comment by dstuecken** [comment]
I just disabled one Test that was always failing, too. The Url where it relies on is not reachable any longer.

**Comment by BernhardPosselt** [comment]
I'd probably put this into a rule file then, e.g.

```php
<?php
return array(
    'grabber' => array(
        '%.*%' => array(
            'test_url' => 'http://www.bbc.co.uk/news/world-middle-east-23911833',
            'body' => array(
                '//div[@class="story-body"]',
            ),
            'after_sucess' => function () ...
        )
    )
);
```

**Comment by BernhardPosselt** [comment]
And regex (your example) is already supported BTW afaik, will check later ;)

**Comment by dstuecken** [comment]
But aren't you hooked on one specific url (e.g. www.bbc.co.uk) with your rule example? Rule files are specified by domain name as far as i can see. I wanted a generic solution no matter which domain the grabber is fetching from.

**Comment by BernhardPosselt** [comment]
I see, hm.

File: tests/Reader/ReaderTest.php [hunk 1]
```diff
@@ -226,9 +226,12 @@ public function testDiscover()
         $this->assertEquals('http://www.universfreebox.com/backend.php', $client->getUrl());
         $this->assertInstanceOf('PicoFeed\Parser\Rss20', $reader->getParser($client->getUrl(), $client->getContent(), $client->getEncoding()));
 
-        $reader = new Reader();
-        $client = $reader->discover('http://planete-jquery.fr');
-        $this->assertInstanceOf('PicoFeed\Parser\Rss10', $reader->getParser($client->getUrl(), $client->getContent(), $client->getEncoding()));
+        /**
+         * http://planete-jquery.fr is currently unreachable and all tests are failing.
```
Comment (fguillot) [inline tests/Reader/ReaderTest.php]: I think you can remove the code instead of commenting.

File: lib/PicoFeed/Processor/ScraperProcessor.php [hunk 2]
```diff
@@ -19,6 +19,24 @@ class ScraperProcessor extends Base implements ItemProcessorInterface
...
+     * @var \Closure
+     */
+    private $executionCallback;
+
+    /**
+     * @param \Closure $executionCallback
+     *
+     * @return $this
+     */
+    public function setExecutionCallback(\Closure $executionCallback)
```
Comment (fguillot) [inline lib/PicoFeed/Processor/ScraperProcessor.php]: Can you add the statement `use Closure;` at the top of file and remove the backslash.

**Comment by fguillot** [comment]
LGTM, can you fix the cosmetic details in the inline comments?

Thanks!

**Comment by fguillot** [comment]
I merged and fixed some coding style.

**Pull request merged**